src/features/broker-protection/actions/build-url-transforms.js

```diff
@@ -0,0 +1,158 @@
+ * These will be applied by default if the key exists in the data.
+ *
+ * @type {Map<string, ((value: string) => string)>}
+ */
+const baseTransforms = new Map([
+    ['firstName', (value) => capitalize(value)],
+    ['lastName', (value) => capitalize(value)],
+    ['state', (value) => value.toLowerCase()],
+    ['city', (value) => capitalize(value)],
+    ['age', (value) => value.toString()]
+])
+
+/**
+ * These are optional transforms, will be applied when key is found in the
+ * variable syntax
+ *
+ * Example, `/a/b/${name|capitalize}` -> applies the `capitalize` transform
+ * to the name field
+ *
+ * @type {Map<string, ((value: string, action: BuildUrlAction) => string)>}
+ */
+const optionalTransforms = new Map([
+    ['hyphenated', (value) => value.split(' ').join('-')],
+    ['capitalize', (value) => capitalize(value)],
+    ['downcase', (value) => value.toLowerCase()],
+    ['upcase', (value) => value.toUpperCase()],
+    ['snakecase', (value) => value.split(' ').join('_')],
+    ['stateFull', (value) => getStateFromAbbreviation(value)],
+    ['ageRange', (value, action) => {
+        if (!action.ageRange) return value
+        const ageNumber = Number(value)
+        // find matching age range
+        const ageRange = action.ageRange.find(range => {
+            const [min, max] = range.split('-')
+            return ageNumber >= Number(min) && ageNumber <= Number(max)
+        })
+        return ageRange || value
+    }]
+])
+
+/**
+ * Take an instance of URLSearchParams and process a new one, with each value
+ * processed
+ *
+ * @param {URLSearchParams} searchParams
+ * @param {BuildUrlAction} action
+ * @param {Record<string, string|number>} userData
+ * @return {URLSearchParams}
+ */
+function processSearchParams (searchParams, action, userData) {
+    /**
+     * For each key/value pair in the URL Search params, process the value
+     * part only.
+     */
+    const updatedPairs = [...searchParams].map(([key, value]) => {
+        const processedValue = processOne(value, action, userData)
+        return [key, processedValue]
+    })
+
+    return new URLSearchParams(updatedPairs)
+}
+
+/**
+ * @param {string} pathname
+ * @param {BuildUrlAction} action
+ * @param {Record<string, string|number>} userData
+ */
+function processPathname (pathname, action, userData) {
+    return pathname
+        .split('/')
+        .filter(Boolean)
+        .map(segment => processOne(segment, action, userData))
+        .join('/')
+}
+
+/**
+ * Process strings like /a/b/${name|lowercase}-${age}
+ * Where the first segment of any variable is the data key, and any
+ * number of subsequent strings are expected to be known transforms
+ *
+ * In that example:
+ *
+ *  - `name` would be processed with the 'lowercase' transform
+ *  - `age` would be used without processing
+ *
+ * @param {string} input
+ * @param {BuildUrlAction} action
+ * @param {Record<string, string|number>} userData
+ */
+export function processOne (input, action, userData) {
```

Comment:
Is this for processing 'one' segment of the URL?

```suggestion
export function processOneURLSegment (input, action, userData) {
```

Reply:
I renamed the function completely to make this clearer :)